src/client/modules/my/dataUpload/dataUpload.ts

```diff
@@ -0,0 +1,127 @@
+                body: JSON.stringify(data)
+            })
+            .then(response => {
+                console.log("Got response: ", response);
+                return response.json();
+            })
+            .then(returnedData => {
+                console.log("Successfully uploaded: ", returnedData);
+            })
+            .catch((error) => {
+                console.log("Error uploading data: ", error);
+            });
+
+        }
+        else {
+
+            this.template.querySelector("p[name=\"errorMessage\"]").hidden = false; //show error message
+
+        }
+
+    }
+
+    /** 
+     * Gets the value from the input field with the given name.
+     * Parameters:
+     *  name: The name of the input field in HTML.
+     * Returns:
+     *  The value currently in the input field. Can be null.
+    */
+    getValueFromInput(name) {
+        return this.template.querySelector("input[name=\"" + name + "\"]").value
+    }
+
+    /**
+     * Validates the input from the form.
+     * Parameters:
+     *  input: The object containing the input data.
+     * Returns:
+     *  True if the data is good to be sent to the database, false otherwise.
+     */
+    validateInput(input) {
+
+        let isDataValid = true; //whether the data is valid
+        let errors = []; //list of error messages
+
+        //check game id
+        if(!input["gameId"]) {
+            errors.push("Game ID cannot be blank.");
+            isDataValid = false;
+        }
+
+        //check each player/victory point pair
+        for(let x = 0; x < input["playerData"].length; x++) {
+
+            //check entries are full
+            if((input["playerData"][x]["playerName"] && !input["playerData"][x]["victoryPoints"]) || 
+                    (!input["playerData"][x]["playerName"] && input["playerData"][x]["victoryPoints"])) {
+
+                errors.push("Non-blank entries must have a player name and a victory point count.");
+                isDataValid = false;
+
+            }
+
+            //check first entry is not empty
+            if(x == 0 && !input["playerData"][x]["playerName"] && !input["playerData"][x]["victoryPoints"]) {
+                errors.push("First entry cannot be blank.");
+                isDataValid = false;
+            }
+
+            //check no entries out of order
+            if(x > 0 && 
+                    (input["playerData"][x]["playerName"] && input["playerData"][x]["victoryPoints"]) && 
+                    (!input["playerData"][x-1]["playerName"] && !input["playerData"][x-1]["victoryPoints"])) {
+
+                errors.push("Please leave no blank rows before entries.");
+                isDataValid = false;
+
+            }
+
+        }
+
+        let errorString = ""; //full error message
+
+        //build error message
+        for(let error of errors)
+            errorString += error + "\n";
+
+        this.template.querySelector("p[name=\"errorMessage\"]").textContent = errorString; //set error text
```

Comment:
Instead of updating the DOM directly, I think it'd be better to return the list of errors from this function. i.e. this becomes a _pure_ function, and the DOM management is left to a more singular place.

Here's a really detailed article if you want to read more: https://medium.com/javascript-scene/master-the-javascript-interview-what-is-a-pure-function-d1c076bec976
This concept is also why I recommend to most that learning Haskell is beneficial, even if no one actually uses it in production.